Refactor SyncthingService (lifecycle), DeviceStateHolder, RestApi, multiple fixes

Catfriend1 opened the pull request:
... SyncthingRunnable, settings UI and listeners.

Fixed issue list from what's already on the tracker:
(fixes #1019) (maybe related to #975)

Fixed issues I discovered testing every UI stress test I could think of and regular syncthing options/wrapper behaviour under various run conditions or foreground run - I created issues for those:
(fixes #1120) - DeviceStateHolder leaks Receivers not unregistered
(fixes #1121) - SettingsActivity calls out to uninitialized RestApi.getLocalDevice() 
(fixes #1122) - NPE in WebGuiActivity.onWebGuiAvailable()
(fixes #1123) - Multiple Web GUI has come online - invalid state transitioning
(fixes #1124) - FirstStartActivity calls startService() - wrong lifecycle
(fixes #1125) - Syncthing is disabled dialog not showing after settingschange
(fixes #1104) - follow-up PR, see post in issue
(fixes #1115) - corresponds to #1104 
(fixes #1126) - Settings screen / run conditions WifiSSIDSelector UI glitch

This PR is a base for further improvement regarding the following issues:
https://github.com/syncthing/syncthing-android/issues/937
https://github.com/syncthing/syncthing-android/issues/914
https://github.com/syncthing/syncthing-android/issues/612
https://github.com/syncthing/syncthing-android/issues/588
https://github.com/syncthing/syncthing-android/issues/523 - as we can make more use of Events API (EventProcessor) in the future after this "spider web" of interacting accross threads/classes and recreating activities more than once got solved.
https://github.com/syncthing/syncthing-android/issues/549 - run conditions can now be further refined

Hi together,
after a 24 hours bug fixing marathon I refactored the syncthing wrapper's code to solve multiple issues related to simultaneously started threads, event processing on listeners, concurrend access to objects that should be locked on multithread access, airplane mode, wifi ssid's settings UI, receiver leaks due to the lack of "isReceiverRegistered" in Android plus registering a receiver more than once or under pref circumstances not registering any receiver which causes syncthing to completely fail startup in background service mode.
I recall something around 20 bugs discovered and fixed by now. I've simplified the code accross classes to a minimum extend of using listeners and callbacks only where necessary. Plus, log output has been added where its crucial to understand the lifecycle of SyncthingService, SyncthingRunnable and asynchronous broadcasts/callbacks from DeviceStateHolder and RestApi. I'll post details and logs on some issues that exist in the current master tomorrow as I'm really tired now (please forgive me). We can then review that whole thing.
I've already tested this to work as much as I could, also stress testing the settings UI and wildly changing real conditions to see if any crash results and for now - it looks good and crash free. The current master has a lot of problems if for example the user stresses the run conditions UI and/or quits syncthing. Also Activity lifecycles weren't correct - now MainActivity is kept on clicking the notification instead of being relaunched against the SingleTask policy (see manifest) All in all, Syncthing wrapper's UI feels much faster and I couldn't get it to crash. Another problem from the master code was, "Syncthing is disabled" wasn't properly maintained or shown if run conditions and/or settings according to that changed or MainActivity was onBackPressed or paused.
I would love your feedback on testing the TeamCity APK.
Promise to open the single issues ASAP so we have a good write up whats fixed by this PR.
Kind regards

Catfriend1 commented:
(draft - still editing)
Fixed issue list from what's already on the tracker:
https://github.com/syncthing/syncthing-android/issues/1019
https://github.com/syncthing/syncthing-android/issues/975
https://github.com/syncthing/syncthing-android/issues/781 - SyncthingService lifecycle now ensures, that states aren't skipped or sent to onApiChange() multiple times.

This PR is a base for further improvement regarding the following issues:
https://github.com/syncthing/syncthing-android/issues/937
https://github.com/syncthing/syncthing-android/issues/914
https://github.com/syncthing/syncthing-android/issues/612
https://github.com/syncthing/syncthing-android/issues/588
https://github.com/syncthing/syncthing-android/issues/523 - as we can make more use of Events API (EventProcessor) in the future after this "spider web" of interacting accross threads/classes and recreating activities more than once got solved.
https://github.com/syncthing/syncthing-android/issues/549 - run conditions can now be further refined

Catfriend1 commented on app/src/main/java/com/nutomic/syncthingandroid/service/ReceiverManager.java:
```diff
@@ -0,0 +1,68 @@
+            return;
+        }
+        Iterator<BroadcastReceiver> iter = mReceivers.iterator();
+        while (iter.hasNext()) {
```
Note - this has to be an iterator and not a foreach(). If using foreach, elements mustn't be removed which we need to do. See foreach() crash that got solved by using the iterator.
![image](https://user-images.githubusercontent.com/16361913/40931469-81367cf8-682b-11e8-9bf0-64aab10506e3.png)

Catfriend1 commented on app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java:
```diff
@@ -140,26 +147,38 @@ public RestApi(Context context, URL url, String apiKey, OnApiAvailableListener a
-    @Override
-    public void onWebGuiAvailable() {
-        mAvailableCount.set(0);
+    public void readConfigFromRestApi() {
```
Relates to https://github.com/syncthing/syncthing-android/issues/1121

Catfriend1 commented on app/src/main/java/com/nutomic/syncthingandroid/activities/WebGuiActivity.java:
```diff
@@ -140,20 +140,31 @@ public void onCreate(Bundle savedInstanceState) {
-            mWebView.stopLoading();
-            setWebViewProxy(mWebView.getContext().getApplicationContext(), "", 0, "localhost|0.0.0.0|127.*|[::1]");
-            mWebView.loadUrl(getService().getWebGuiUrl().toString());
+    public void onServiceStateChange(SyncthingService.State newState) {
```
Fix for https://github.com/syncthing/syncthing-android/issues/1122 - Solution:
Only expose onWebGuiAvailable to RestApi (where the only important place from my point of view is where it's really needed) and implement onServiceStateChange() (before PR it's called onApiChange)

Catfriend1 commented:
@AudriusButkevicius Ready for review. I can confirm this build is working perfectly on my Android 7.1.2 phone model lg-h815. I tested the shit out of it as much as I could. Would appreciate more people testing the debug apk and giving feedback. All in all, syncthingService now sticks to clear lifecycle DISABLED > INIT > STARTING > ACTIVE > DISABLED > ... (as the most used lifeycle). This is the main achievement making the code  as @Nutomic said beforehand "it could be in the need of a complete rewrite" more straightforward eliminating easy to discover bugs and settling sophisticated threadsafe operation correct.

AudriusButkevicius commented:
Uhh, this is unreviewable.
Can we break this down into 5 or however many we need small PRs that address one issue at a time?
There is so much stuff moving here that it's impossible to grasp, and the best I could do is just merge with my eyes closed, and hope for the best.

AudriusButkevicius commented:
`#1121 - SettingsActivity calls out to uninitialized RestApi.getLocalDevice()`
I know Google Play console is full on NPEs related to the RestApi being uninitialized, so we should have a closer look everywhere it's used.

Catfriend1 commented:
About those restApi NPE's you are right. They are caused because of the broken lifecycle the syncthingservice goes through especially on run condition change. The most fundamental fault I discovered was signalling State.Active where other modules expect restapi to be fully ready called out a) too early by bad implementation of the callback or b) staggering concurrently running pollers calling back multiple times causing multiple concurrent accessing of restapi init stuff and variables.

Catfriend1 commented:
I know this is hard to review as I wanted to do the code refactoring initially focusing on two issues but everytime trying to incrementally verify it works correctly a new fundamental Problem with the code came up. To get it quality assured I put a log more verbose log in the "hot" places and played over my whole stress test and real test case procedure again and again last night after making improvements until no more issues were found aroung the wrappers core classes. Do you think merging as a separate beta tag and release is possible? Wouls you mind in assisting me by trying out a test build? I've also had my difficulties with all the logs from yesterday creating those issues and referring to what part was mostly relevant in fixing them. Split up would be very time consuming and - having experienced how long the rewrite of certain parts and dependencies took due to a lot of thread, listener and callback code - took, most likely from my point of view not give us an stable master build.

AudriusButkevicius commented:
You should really try to work in an organized way.
I know it's easy to get side-tracked and go fix all the bugs you find, on the left and right, but remember that someone will have to understand the change and approve it. If you see something easy to fix, `git add -u; git commit -m WIP; git checkout origin/master; git branch fix-this-other-thing` and fix that small thing and file a separate PR.

When contributing to Go, I was asked to open a separate PR for renaming a variable, so that renames and functionality lives in separate PRs.
This is just to set the gauge of the atmosphere when 1k blob of changes comes in, with convoluted ideas, and no idea which change fixes what issue or why it's there to start with.

I am pretty sure no open source project would appreciate this type of pull request, even if it fixes all bugs under the sun, especially as there is no tests proving that it actually works.

I don't use syncthing, so yeah, I can start the app and see that it launches, but this is as far as it goes me helping test things. 

Sure, we can make a gradual rollout of the next version, and **hope** that people catch the issues before they figure out that we shat in their bed and come for us with pitchforks.

AudriusButkevicius commented:
I mean, I made some meaningless comments, as I have no idea why some of the changes are happening in the first place.

I guess once you deal with my "it's named wrong, I think you need a lock" nonsense, we'll let it brew for a while in hopes that someone else comes in and reviews this.

If not, we'll merge it, and hope for the best.
If shit does hit the fan, I'll let you do the cleanup I guess.

Catfriend1 commented:
Thanks for taking time to comment. I'm looking into every comment tomorrow and answer as good as I can. Also found some parts I can improve by your feedback before we discuss merge. I'm sleeping a night over it if I can split it up, maybe on module base. Youre right that this was an uncommon way to get from one thing to another rewriting more and more... Hope you'll forgive me finally in the favour of a fixed basement the app stands on. I definitely don't intend to work like this untargeted in the future. If issues arise from the changes I'll accept the work to solve them. Getting back with answers shortly...

Catfriend1 commented:
@AudriusButkevicius  I took care of the review aspects. Any more suggestions to improve this PR? If not, please leave me a message so I will do a time expensive testing of the whole app with a) a new setup and b) an existing setup including different settings/settings and run condition changes/stress testing, large and small databases to verify everything works correctly before merge. If something breaks, I'll assist in fixing them, sure.

Catfriend1 commented:
Update: I roughly tested the current debug build on latest commit after the big test yesterday and it works without a single android runtime exception. Done this rough test earlier than planned to decide if I can point a ticket OP to the newest debug build. Lifecycles are okay, receivers too. What I did:
Installed debug apk fresh
Stress tested ui with run condition changes and changes to prefs
Imported my sync config
Restarted syncthing app from the drawer menu
Opened web gui activity
Clicked the status bar notification after switching away.
Went to settings and set conditions so syncthing wont run, got back to main activity, hit exit, clicked the notification, changed prefs back for running syncthing
Opened the web ui
Exited the app.
Made some nice looking photos of the logs, logs are okay, will upload them later.

P.s. Should I put all fixed issues (as they are many) in the PR's title?

AudriusButkevicius commented:
Put them in the body, I think git parses that too.

Catfriend1 commented:
@AudriusButkevicius Done, I also put my log screenshots up for proving this to be verified and working on my Android 7.1.2 phone.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java:
```diff
@@ -124,11 +142,6 @@ public RestApi(Context context, URL url, String apiKey, OnApiAvailableListener a
-
     public interface OnApiAvailableListener {
         void onApiAvailable();
     }
```
Is this still used?

AudriusButkevicius replied:
The interface, is that used?

Catfriend1 replied:
| The interface, is that used? |
Did this overlap with my last post? > Answer is yes, see SynctingService#onApiAvailable

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/RestApi.java:
```diff
@@ -311,7 +330,13 @@ private void removeFolderInternal(String id) {
+        List<Device> devices = getDevices(true);
+        if (devices.isEmpty()) {
+            Log.e(TAG, "getLocalDevice: devices is empty");
+            throw new RuntimeException();
```
That's not a very useful exception...

AudriusButkevicius replied:
That's still not a useful exception, regardless that the error message is good.
Error messages will not be sent to google play.

Catfriend1 replied:
Ok got the point. So we need a good error message thrown "to google play". I'm going to improve this shorty ...

Catfriend1 replied:
I put the log message to be thrown as RuntimeException. commit https://github.com/syncthing/syncthing-android/pull/1119/commits/9bdc1a7ebf72bd563bf2e7b2340479db9ead5d97

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingRunnable.java:
```diff
@@ -263,48 +263,43 @@ private boolean useWakeLock() {
+            if (ret != 0) {
+                Log.e(TAG_NICE, "Failed to set ionice " + Integer.toString(ret));
+            }
+        }
```
Why was this in a thread before?

Catfriend1 replied:
Removing it was just a proposal as I think the person getting this in meant to address an older syncthing version before 0.14.47. I'm reverting this as I agree with you we can remove it later. To sum up my answer to the question: No need for a new thread. Can stay as it is and won't do harm.
Revert with commit https://github.com/syncthing/syncthing-android/pull/1119/commits/bd08037e07e81c37d54cca5a4e6e30ba3d62e880

AudriusButkevicius replied:
`"nices out itself" on since binary version 0.14.47+`, not ionice, which is what this is doing.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -59,18 +59,8 @@
-
-    public interface OnApiChangeListener {
-        void onApiChange(State currentState);
+    public interface onServiceStateChangeListener {
```
Interface should start with a capital letter. I am surprised lint does not bitch about it.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/activities/StateDialogActivity.java:
```diff
@@ -69,24 +79,25 @@ private void onApiChange(SyncthingService.State currentState) {
-                dismissLoadingDialog();
-                if (!isFinishing()) {
+                if (!mIsPaused) {
                     showDisabledDialog();
```
Will this not show a dialog on top?

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/DeviceStateHolder.java:
```diff
@@ -70,19 +71,26 @@ public static boolean alwaysRunInBackground(Context context) {
+    // Those receivers are managed by {@link mReceiverManager}.
+    private NetworkReceiver mNetworkReceiver;
+    private BatteryReceiver mBatteryReceiver;
+    private PowerSaveModeChangedReceiver mPowerSaveModeChangedReceiver = null;
```
Why is this nulled, and others are not?

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -234,34 +274,86 @@ protected Void doInBackground(Void... voids) {
-            pollWebGui();
+            // Start the syncthing binary.
+            if (mSyncthingRunnable != null || mSyncthingRunnableThread != null) {
+                Log.e(TAG, "StartupTask/onPostExecute: Syncthing binary lifecycle violated");
```
This errors does no good. Should we crash or just return here?

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -234,34 +274,86 @@ protected Void doInBackground(Void... voids) {
+        if (mEventProcessor == null) {
+            mEventProcessor = new EventProcessor(SyncthingService.this, mApi);
+        }
+        mEventProcessor.start();
```
Do you really want to double start it if it already exists?

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/service/SyncthingService.java:
```diff
@@ -350,60 +451,37 @@ public void registerOnWebGuiAvailableListener(OnWebGuiAvailableListener listener
-            mOnWebGuiAvailableListeners.clear();
-        });
+    public void unregisterOnServiceStateChangeListener(onServiceStateChangeListener listener) {
+        mOnServiceStateChangeListeners .remove(listener);
```
space before dot.

AudriusButkevicius commented:
So the screenshots of logs have literally no value to me, just makes the page harder to load.

I made some new comments, and a lot of my questions are still not answered.

AudriusButkevicius commented:
I ask a question and you start removing things.
I ask to put a message in the exception, you expand the message in log line, that is never ever visible to anyone.

Are you reading the review comments or are you still working on this and I should come back later?

Catfriend1 commented:
@AudriusButkevicius Yeah the logs make more sense if one debugs the app with multiple test plans manually, collects logs before and after this PR and compares them to see where improvements act a new way.
Sorry that I've overseen 2 or 3 comments unintentionally, I didn't notice they were hidden behind "Load diff" looking through your review. I answered them plus the new ones and made improvement commits :). Thanks for reviewing this big a second time.
Are we ready now for me doing the "big test" of the latest debug apk again to verify this PR working?

AudriusButkevicius commented:
There are still plenty of things that were not addressed/answered.

AudriusButkevicius commented:
Also removing the ionice and stuff? Why are we removing it? If there is a good reason, why does this have to happen in this PR?

Catfriend1 commented:
@AudriusButkevicius 
I've now answered and addressed everything that was on the code review (link https://github.com/syncthing/syncthing-android/pull/1119/files/872e7721ea989c2750ab99555f249fde3a2a25bb ) did I miss something?
At the time of you writing I'd have to apologize that I did not F5 this page working on the review tab and got this afterwards.  I was indeed still going to the review on doing improvements that came up to my mind reading your questions. Now I'm done. The RuntimeException was a misunderstanding as I just focused on the logs, it's improved now. Thanks for hitting my head hard to the wall until I finally got the point :).
I reverted the niceSyncthing removal so we have this part in again, agreed to remove it later as this is really big now.

AudriusButkevicius commented:
There is still plenty of stuff unanswered, just go to the files tab, and check.

Catfriend1 commented:
Sorry, I logged out, clicked the link and recognized you couldn't see my comments?! I was somehow in pending review mode and hit this button
![image](https://user-images.githubusercontent.com/16361913/41006287-2de77eca-6921-11e8-8b3e-e7adb5e9b2d2.png)
total github beginner :(

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/fragments/DrawerFragment.java:
```diff
@@ -136,11 +136,16 @@ public void onSaveInstanceState(Bundle outState) {
      */
     private void updateGui() {
-        if (mActivity.getApi() == null || getActivity() == null || getActivity().isFinishing())
+        if (getActivity() == null || getActivity().isFinishing()) {
```
There is a race between checking if getActivity is null and it actually being null on the secons call, causing a NPE. Probably store the getActivitiy value in a variable.

Catfriend1 replied:
That's better, done with commit https://github.com/Catfriend1/syncthing-android/commit/13a180dc49c27bf80a3d1565f55dd55e93d96a0c

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/activities/DeviceActivity.java:
```diff
@@ -214,7 +214,7 @@ private void restoreDialogStates(Bundle savedInstanceState) {
         super.onDestroy();
         if (getService() != null) {
-            getService().unregisterOnApiChangeListener(this::onApiChange);
+            getService().unregisterOnServiceStateChangeListener(this::onServiceStateChange);
```
What if getServics returns a null on the second invocation?

Catfriend1 replied:
improved with commit https://github.com/Catfriend1/syncthing-android/commit/66ed4bdda02820267bd1dc33b31561a8eac1b54e
This is just a safety check already present on the old code. But yes, I agree storing the ref on a variable and use this is better so I did improve that. As we shutdown the syncthing service first an d then finish MainActivity (holding this fragment) which calls onDestroy() getService couldn't get null while this code part is running at the moment (but in a future implementation maybe :) )

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/activities/FolderActivity.java:
```diff
@@ -235,7 +235,7 @@ private Bundle getVersioningBundle() {
         super.onDestroy();
         if (getService() != null) {
-            getService().unregisterOnApiChangeListener(this);
+            getService().unregisterOnServiceStateChangeListener(this);
```
Ditto about second call

Catfriend1 replied:
improved with commit https://github.com/Catfriend1/syncthing-android/commit/66ed4bdda02820267bd1dc33b31561a8eac1b54e
see above.

AudriusButkevicius commented on app/src/main/java/com/nutomic/syncthingandroid/activities/FolderPickerActivity.java:
```diff
@@ -156,13 +156,13 @@ private void populateRoots() {
         super.onDestroy();
-        getService().unregisterOnApiChangeListener(this);
+        getService().unregisterOnServiceStateChangeListener(this);
     }
```
Can this be null?

Catfriend1 replied:
No can't be, same explanation as I posted in https://github.com/syncthing/syncthing-android/pull/1119#discussion_r193385835 (see DeviceActivity question).
We should anyway null check this to make future implementations more easy not to run into issues here. Done by commit https://github.com/Catfriend1/syncthing-android/commit/9393e6fdd8869c93a95ad16fb76bb1fcd7b3ba0b .

AudriusButkevicius commented:
I guess I already made it quite clear previously, but, there is no way anything like this will ever get merged again.

Catfriend1 commented:
Right, okay. I'm also standing for this ugly pr to be a one time occurence to make the code clean.

Catfriend1 commented:
Note on testing: I got 2h 53m stable syncthing binary execution testing with Android 7.1.2 on lg-h815 with the RestartOnWakeup option set to disabled.
![image](https://user-images.githubusercontent.com/16361913/41141969-5b56ddfe-6af4-11e8-921f-9a8464b2b2dc.png)

capi commented:
Sorry, I will not have time to really review this huge PR (thanks for the gigantic work, @Catfriend1), and I read along in the mails with @AudriusButkevicius doing a great job at reviewing the changes while they came in.

I'll try to compile and install this branch sometime this weekend and just try if everything _in my usecase_ works, in order to contribute a little bit to testing.

I am with @AudriusButkevicius that PRs should be smaller. For this huge PR I am struggling a bit if we should merge it squashed or not? Normally I am pro-squashing on merge, but the individual steps at least show the intention of most changes. But having a huge block of 83 commits merged in, I'm just not sure.

Catfriend1 commented:
@capi : If you need the latest commit of this PR built, you can grab it here https://build.syncthing.net/repository/download/SyncthingAndroid_Build/22103:id/debug/app-debug.apk
I suggest squashing commits from beginning until (https://github.com/syncthing/syncthing-android/pull/1119/commits/6c00e7aad0b93dcd70daa9bed90c3aa0735a75f3 ) together into one commit. After this "WIP" commit, every commit addressed one issue at once not affecting the whole bunch of code that much then it was before.

capi commented:
Thanks, but I need to compile myself, since my current version is self-compiled and installed and otherwise the signing keys won't match.

Catfriend1 commented:
@capi : I think you have your reasons for the self-compile :) let me know if you have some code ready we can check in perhaps for everyone to benefit or make the app's syncthing data backup & restore more mightful. (I recently thought about also backing up the index but for unexperienced users this could be risky)
